File: androidutils/src/main/java/com/uwetrottmann/androidutils/AndroidUtils.java
package com.uwetrottmann.androidutils;

import android.Manifest;
import android.annotation.TargetApi;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Build;
import android.os.Environment;
import android.text.TextUtils;
import android.view.View;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.channels.FileChannel;
import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresPermission;

/**
 * Provides various static helper methods to
 * <ul>
 * <li>check for Android version (like {@link #isAtLeastOreo()}),</li>
 * <li>an active or unmetered network connection,</li>
 * <li>external storage state availability.</li>
 * </ul>
 */
public class AndroidUtils {

    /*
     * NOTE: new Android SDK version checks should contain the code name,
     * otherwise Lint does not seem to detect them?!
     */

    /**
     * API level 30+, Android 11.
     */
    public static boolean isAtLeastR() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.R;
    }

    /**
     * API level 29+, Android 10.
     */
    public static boolean isAtLeastQ() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q;
    }

    /**
     * API level 28+, Android 9.
     */
    public static boolean isAtLeastPie() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.P;
    }

    /**
     * API level 27+, Android 8.1.
     */
    public static boolean isAtLeastOreoMR1() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1;
    }

    /**
     * API level 26+, Android 8.0.
     */
    public static boolean isAtLeastOreo() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.O;
    }

        /**
         * API level 25+, Android 7.1.
         */
    public static boolean isNougatMR1OrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1;
    }

    /**
     * API level 24+, Android 7.0.
     */
    public static boolean isNougatOrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.N;
    }

    /**
     * API level 23+, Android 6.0.
     */
    public static boolean isMarshmallowOrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.M;
    }

    /**
     * API level 22+, Android 5.1.
     */
    public static boolean isLollipopMR1OrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP_MR1;
    }

    /**
     * API level 21+, Android 5.0.
     */
    public static boolean isLollipopOrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP;
    }

    /**
     * API level 20+, Android 4.4.
     */
    public static boolean isKitKatWatchOrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT_WATCH;
    }

    /**
     * API level 19+, Android 4.4.
     */
    public static boolean isKitKatOrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT;
    }

    /**
     * API level 18+, Android 4.3.
     */
    public static boolean isJellyBeanMR2OrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2;
    }

    /**
     * API level 17+, Android 4.2.
     */
    public static boolean isJellyBeanMR1OrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1;
    }

    /**
     * API level 16+, Android 4.1.
     */
    public static boolean isJellyBeanOrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN;
    }

    /**
     * API level 15+, Android 4.0.3.
     */
    public static boolean isICSMR1OrHigher() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.ICE_CREAM_SANDWICH_MR1;
    }

    /**
     * Checks if {@link Environment}.MEDIA_MOUNTED is returned by {@code getExternalStorageState()}
     * and therefore external storage is read- and writeable.
     */
    public static boolean isExtStorageAvailable() {
        return Environment.MEDIA_MOUNTED.equals(Environment.getExternalStorageState());
    }

    /**
     * Returns true if the current layout direction is {@link View#LAYOUT_DIRECTION_RTL}.
     *
     * <p>This always returns false on versions below JELLY_BEAN_MR1.
     */
    public static boolean isRtlLayout() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            int direction = TextUtils.getLayoutDirectionFromLocale(Locale.getDefault());
            return direction == View.LAYOUT_DIRECTION_RTL;
        }
        return false;
    }

    @Nullable
    private static ConnectivityManager getConnectivityManager(@NonNull Context context) {
        // use application context to avoid leaking any activity
        // https://code.google.com/p/android/issues/detail?id=198852
        return (ConnectivityManager) context.getApplicationContext()
                .getSystemService(Context.CONNECTIVITY_SERVICE);
    }

    @RequiresPermission(Manifest.permission.ACCESS_NETWORK_STATE)
    @Nullable
    private static NetworkInfo getActiveNetworkInfo(@NonNull Context context) {
        ConnectivityManager connectivityManager = getConnectivityManager(context);
        if (connectivityManager == null) {
            return null;
        }
        return connectivityManager.getActiveNetworkInfo();
    }

    /**
     * Whether there is an active network connection.
     */
    @RequiresPermission(Manifest.permission.ACCESS_NETWORK_STATE)
    public static boolean isNetworkConnected(@NonNull Context context) {
        NetworkInfo activeNetworkInfo = getActiveNetworkInfo(context);
        return activeNetworkInfo != null && activeNetworkInfo.isConnected();
    }

    /**
     * Whether there is an active network connection and it is via WiFi.
     *
     * <p>If you want to check whether to transmit large amounts of data, you may want to use {@link
     * #isUnmeteredNetworkConnected(Context)}.
     */
    @RequiresPermission(Manifest.permission.ACCESS_NETWORK_STATE)
    public static boolean isWifiConnected(@NonNull Context context) {
        NetworkInfo activeNetwork = getActiveNetworkInfo(context);
        return activeNetwork != null && activeNetwork.isConnected()
                && activeNetwork.getType() == ConnectivityManager.TYPE_WIFI;
    }

    /**
     * Whether there is an active network connection and it is not metered, e.g. so large amounts of
     * data may be transmitted.
     */
    @RequiresPermission(Manifest.permission.ACCESS_NETWORK_STATE)
    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    public static boolean isUnmeteredNetworkConnected(@NonNull Context context) {
        ConnectivityManager connectivityManager = getConnectivityManager(context);
        if (connectivityManager == null) {
            return false;
        }
        NetworkInfo activeNetworkInfo = connectivityManager.getActiveNetworkInfo();
        return activeNetworkInfo != null && activeNetworkInfo.isConnected()
                && !connectivityManager.isActiveNetworkMetered();
    }

    /**
     * Copies the contents of one file to the other using {@link FileChannel}s.
     *
     * @param src source {@link File}
     * @param dst destination {@link File}
     */
    public static void copyFile(@NonNull File src, @NonNull  File dst) throws IOException {
        FileInputStream in = new FileInputStream(src);
        FileOutputStream out = new FileOutputStream(dst);
        FileChannel inChannel = in.getChannel();
        FileChannel outChannel = out.getChannel();
        try {
            inChannel.transferTo(0, inChannel.size(), outChannel);
        } finally {
            try {
                in.close();
            } catch (IOException ignored) {
            }
            try {
                out.close();
            } catch (IOException ignored) {
            }
        }
    }
}

File: androidutils/src/main/java/com/uwetrottmann/androidutils/AndroidUtils.kt
package com.uwetrottmann.androidutils

import android.Manifest
import android.annotation.TargetApi
import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import android.os.Environment
import android.text.TextUtils
import android.view.View
import androidx.annotation.ChecksSdkIntAtLeast
import androidx.annotation.RequiresApi
import androidx.annotation.RequiresPermission
import androidx.core.content.getSystemService
import com.uwetrottmann.androidutils.AndroidUtils.isAtLeastOreo
import java.util.*

/**
 * Provides various static helper methods to
 *
 *  * check for Android version (like [isAtLeastOreo]),
 *  * an active or unmetered network connection,
 *  * external storage state availability.
 *
 */
@Suppress("unused")
object AndroidUtils {

    /**
     * API level 31+, Android 12.
     */
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES.S)
    @JvmStatic
    val isAtLeastS: Boolean
        get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.S

    /**
     * API level 30+, Android 11.
     */
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES.R)
    @JvmStatic
    val isAtLeastR: Boolean
        get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.R

    /**
     * API level 29+, Android 10.
     */
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES.Q)
    @JvmStatic
    val isAtLeastQ: Boolean
        get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q

    /**
     * API level 28+, Android 9.
     */
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES.P)
    @JvmStatic
    val isAtLeastPie: Boolean
        get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.P

    /**
     * API level 27+, Android 8.1.
     */
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES.O_MR1)
    @JvmStatic
    val isAtLeastOreoMR1: Boolean
        get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1

    /**
     * API level 26+, Android 8.0.
     */
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES.O)
    @JvmStatic
    val isAtLeastOreo: Boolean
        get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.O

    /**
     * API level 25+, Android 7.1.
     */
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES.N_MR1)
    @JvmStatic
    val isNougatMR1OrHigher: Boolean
        get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1

    /**
     * API level 24+, Android 7.0.
     */
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES.N)
    @JvmStatic
    val isNougatOrHigher: Boolean
        get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.N

    /**
     * API level 23+, Android 6.0.
     */
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES.M)
    @JvmStatic
    val isMarshmallowOrHigher: Boolean
        get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.M

    /**
     * API level 22+, Android 5.1.
     */
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES.LOLLIPOP_MR1)
    @JvmStatic
    val isLollipopMR1OrHigher: Boolean
        get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP_MR1

    /**
     * API level 21+, Android 5.0.
     */
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES.LOLLIPOP)
    @JvmStatic
    val isLollipopOrHigher: Boolean
        get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP

    /**
     * API level 20+, Android 4.4.
     */
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES.KITKAT_WATCH)
    @JvmStatic
    val isKitKatWatchOrHigher: Boolean
        get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT_WATCH

    /**
     * API level 19+, Android 4.4.
     */
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES.KITKAT)
    @JvmStatic
    val isKitKatOrHigher: Boolean
        get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT

    /**
     * API level 18+, Android 4.3.
     */
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES.JELLY_BEAN_MR2)
    @JvmStatic
    val isJellyBeanMR2OrHigher: Boolean
        get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2

    /**
     * API level 17+, Android 4.2.
     */
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES.JELLY_BEAN_MR1)
    @JvmStatic
    val isJellyBeanMR1OrHigher: Boolean
        get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1

    /**
     * API level 16+, Android 4.1.
     */
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES.JELLY_BEAN)
    @JvmStatic
    val isJellyBeanOrHigher: Boolean
        get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN

    /**
     * Checks if [Environment].MEDIA_MOUNTED is returned by `getExternalStorageState()`
     * and therefore external storage is read- and writeable.
     */
    @JvmStatic
    val isExtStorageAvailable: Boolean
        get() = Environment.MEDIA_MOUNTED == Environment.getExternalStorageState()

    /**
     * Returns true if the current layout direction is [View.LAYOUT_DIRECTION_RTL].
     *
     * This always returns false on versions below JELLY_BEAN_MR1.
     */
    @JvmStatic
    val isRtlLayout: Boolean
        get() {
            if (isJellyBeanMR1OrHigher) {
                val direction = TextUtils.getLayoutDirectionFromLocale(Locale.getDefault())
                return direction == View.LAYOUT_DIRECTION_RTL
            }
            return false
        }

    private fun getConnectivityManager(context: Context): ConnectivityManager? {
        // use application context to avoid leaking any activity
        // https://code.google.com/p/android/issues/detail?id=198852
        return context.applicationContext.getSystemService()
    }

    /**
     * Uses Android 6 APIs to get active network capabilities.
     *
     * Note: due to a race condition this API can return a SecurityException for the lifetime of
     * the app on Android 11 devices that did not [receive a patch](https://android-review.googlesource.com/c/platform/frameworks/base/+/1758029),
     * see [reported issue](https://issuetracker.google.com/issues/175055271).
     *
     * https://developer.android.com/training/basics/network-ops/reading-network-state
     */
    @RequiresApi(Build.VERSION_CODES.M)
    @RequiresPermission(Manifest.permission.ACCESS_NETWORK_STATE)
    fun getActiveNetworkCapabilities(context: Context): NetworkCapabilities? {
        val connectivityManager = getConnectivityManager(context) ?: return null
        val activeNetwork = connectivityManager.activeNetwork ?: return null
        return connectivityManager.getNetworkCapabilities(activeNetwork)
    }

    /**
     * Whether the active network has [NetworkCapabilities.NET_CAPABILITY_INTERNET],
     * or on Android 5.1 or lower if the active network connection is connected.
     *
     * Note: on Android 11 this also uses the old APIs, see [getActiveNetworkCapabilities].
     */
    @JvmStatic
    @RequiresPermission(Manifest.permission.ACCESS_NETWORK_STATE)
    fun isNetworkConnected(context: Context): Boolean {
        if (isMarshmallowOrHigher && Build.VERSION.SDK_INT != Build.VERSION_CODES.R) {
            val networkCapabilities = getActiveNetworkCapabilities(context) ?: return false
            return networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
        } else {
            val connectivityManager = getConnectivityManager(context) ?: return false
            @Suppress("DEPRECATION")
            val activeNetworkInfo = connectivityManager.activeNetworkInfo
            @Suppress("DEPRECATION")
            return activeNetworkInfo != null && activeNetworkInfo.isConnected
        }
    }

    /**
     * Whether the active network has [NetworkCapabilities.NET_CAPABILITY_INTERNET] and
     * [NetworkCapabilities.NET_CAPABILITY_NOT_METERED], or on Android 5.1 or lower if the active
     * network connection is connected and not metered.
     *
     * This can be used to determine if e.g. large amounts of data may be transmitted.
     *
     * Note: on Android 11 this also uses the old APIs, see [getActiveNetworkCapabilities].
     */
    @JvmStatic
    @RequiresPermission(Manifest.permission.ACCESS_NETWORK_STATE)
    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    fun isUnmeteredNetworkConnected(context: Context): Boolean {
        if (isMarshmallowOrHigher && Build.VERSION.SDK_INT != Build.VERSION_CODES.R) {
            val networkCapabilities = getActiveNetworkCapabilities(context) ?: return false
            return networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
                    && networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_METERED)
        } else {
            val connectivityManager = getConnectivityManager(context) ?: return false
            @Suppress("DEPRECATION")
            val activeNetworkInfo = connectivityManager.activeNetworkInfo
            @Suppress("DEPRECATION")
            return (activeNetworkInfo != null && activeNetworkInfo.isConnected
                    && !connectivityManager.isActiveNetworkMetered)
        }
    }

}